_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,359 @@
+
+Unfortunately, the idempotent producer does not guarantee duplicate protection across restarts and when you need to write to multiple partitions as a single unit of work.
+This is the typical scenario with many stream processing applications that run read-process-write cycles.
+In this case, you can make these cycles atomic using the transaction support of the Kafka Producer or Kafka Streams APIs.
+
+The transaction support exposed by the Producer API is low level, and needs to be used with carefully in order to actually get transactional semantics.
+For those using Kafka Streams it is much simpler: setting `processing.guarantee=exactly_once_v2` will enable EOS on your existing topology.
+
+### Considerations
+
+Each producer must configure its own static and unique [`transactional.id`](https://kafka.apache.org/documentation/#producerconfigs_transactional.id) to enable the zombie fencing logic, which avoids message duplicates.
+
+The `transactional.id` is used to uniquely identify the same logical producer across process restarts.
+In contrast to the idempotent producer, a transactional producer instance will be allocated the same PID (but incremented producer epoch), as any previous instance with the same `transactional.id`. 
+Together the PID and producer epoch identify a logical producer session.
+
+When starting, the producer call `initTransactions` to initialize the producer session, allowing the broker to tidy up state associated with any previous incarnations of this producer, as identified by the `transactional.id`.
+Thereafter, any existing producers using the same `transactional.id` are considered zombies, and are fenced off as soon as they try to send data.
+
+> Before Kafka 2.6 the `transactional.id` had to be a static encoding of the input partitions in order to avoid ownership transfer between application instances during rebalances, that would invalidate the fencing logic.
+> This limitation was hugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition. 
+> It was fixed with by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).
+
+To guarantee message ordering, a given producer can have at most one open transaction (they are executed serially).
+Consumers with `isolation.level=read_committed` are not allowed to advance to offsets which include open transactions, while messages from aborted transactions are filtered out.
+
+The transaction coordinator is a module running inside a Kafka broker.
+Each coordinator owns a subset of the partitions in the internal `__transaction_state` topic (i.e. the partitions for which its broker is the leader).
```

Comment:
I don't really understand the point of this paragraph. Why are we telling the reader about the transaction coordinator _here_? It might make more sense to introduce it when you talk about the PID for transactional producer.

Reply:
Moved, thanks.